### PR TITLE
Gossiper: get_or_create_endpoint_state update this node RPC_ADDRESS and HOST_ID

### PR DESCRIPTION
**Pull request by bhalevy**
When setting this node's RPC_ADDRESS application state,
use the `broadcast_rpc_address` rather than `broadcast_address`
as this is the one that should be exposed to the client.
See also `storage_service::join_token_ring()`.

Also, set the endpoint_state HOST_ID if available.

Fixes scylladb/scylladb#15458

### CODE REVIEW

File: gms/gossiper.cc [hunk 1]
```diff
@@ -1428,7 +1428,10 @@ endpoint_state& gossiper::get_or_create_endpoint_state(inet_address ep) {
     auto it = _endpoint_state_map.find(ep);
     if (it == _endpoint_state_map.end()) {
         auto eps = endpoint_state();
-        eps.add_application_state(application_state::RPC_ADDRESS, versioned_value::rpcaddress(ep));
+        if (utils::fb_utilities::is_me(ep)) {
+            auto rpc_addr = utils::fb_utilities::get_broadcast_rpc_address();
+            eps.add_application_state(application_state::RPC_ADDRESS, versioned_value::rpcaddress(rpc_addr));
```
Comment (tgrabiec) [inline gms/gossiper.cc:1433]: Why do we set RPC_ADDRESS here? Can we create an empty endpoint_state?

File: gms/gossiper.cc [hunk 2]
```diff
@@ -1431,6 +1431,10 @@ endpoint_state& gossiper::get_or_create_endpoint_state(inet_address ep) {
         if (utils::fb_utilities::is_me(ep)) {
             auto rpc_addr = utils::fb_utilities::get_broadcast_rpc_address();
             eps.add_application_state(application_state::RPC_ADDRESS, versioned_value::rpcaddress(rpc_addr));
+            auto node = get_token_metadata_ptr()->get_topology().this_node();
+            if (node && node->host_id()) {
+                eps.add_application_state(application_state::HOST_ID, versioned_value::host_id(node->host_id()));
```
Comment (tgrabiec) [inline gms/gossiper.cc:1436]: What's the reason for this change?

Comment (bhalevy) [reply gms/gossiper.cc:1436]: prep for https://github.com/scylladb/scylladb/issues/12282

File: gms/gossiper.cc (hunk 1 above)
Comment (tgrabiec) [reply gms/gossiper.cc:1433]: It seems arbitrary that we initialize just this state here. There is much more states, they're added in storage_service.

Some states have dependencies between them, and they should appear together. It would be incorrect to add them at different points in time, and allow other nodes to observe inconsistent set of states. So why not leave adding all states to where it is done now, in storage service?

File: gms/gossiper.cc (hunk 2 above)
Comment (tgrabiec) [reply gms/gossiper.cc:1436]: But you don't have to add HOST_ID to create an endpoint state for a given host_id.

File: gms/gossiper.cc (hunk 1 above)
Comment (bhalevy) [reply gms/gossiper.cc:1433]: I'm OK with making this change but what inter-dependencies the RPC_ADDRESS state have?

File: gms/gossiper.cc (hunk 2 above)
Comment (bhalevy) [reply gms/gossiper.cc:1436]: Why complicate our lives?
If the endpoint_state is guaranteed to contain the HOST_ID it is self-identifyable.

Comment (bhalevy) [reply gms/gossiper.cc:1436]: Maybe it will make more sense in the context of #15444

**Comment by bhalevy** [comment]
Closing this PR in favor of #15475
